### PR TITLE
feat: support running extensions on private code without a private Sourcegraph instance

### PR DESCRIPTION
**Pull request by chrismwendt**
**Motivation** This adds support for running extensions (e.g. Codecov) on private code even when the repository does not exist on the Sourcegraph instance. Doing so will resolve this issue: https://github.com/sourcegraph/browser-extensions/issues/234#issuecomment-430429015

**Problem** The browser extension must not let arbitrary GraphQL requests through to the Sourcegraph instance (which could be Sourcegraph.com) because they might contain private info (such as repo names), but some GraphQL requests are necessary to run extensions at all (e.g. fetching the extension manifests).

**Implementation** I've marked some GraphQL requests as safe by passing in `requestMightContainPrivateInfo: false` (defaults to `true` to keep old behavior for unmarked requests). The browser extension will allow those requests to be sent to the Sourcegraph instance. I also blocked `queryLSP` when sent from a private repo to Sourcegraph.com.

**Reviewers** The main file to review is `src/shared/backend/extensions.ts`. If you want to review the rest, I'd recommend reviewing commit-by-commit.

- @ijsnow I'd like your :eyes: on "refactor: avoid a network request while resolving the revision"
- @sqs I'd like your input on the interaction between e-c-c and the browser extension in "feat: allow GraphQL requests that are known to not contain private info" and "feat: block queryLSP from a private repo to Sourcegraph.com" and whether or not this approach makes sense

**Testing**

- `git checkout allow-known-safe-gql-requests`
- `yarn ; yarn run dev`
- Make sure one language extension (e.g. Go) and one other extension (e.g. token-highlights) have the following behavior in these scenarios:
  - Private repository that doesn't exist on a private Sourcegraph instance: token-highlights should work, Go should not (also, Go should not spam you with 404s in the hover tooltips)
  - Private repository that exists on a private Sourcegraph instance: both should work
  - Private repository that doesn't exist on Sourcegraph.com: token-highlights should work, Go should not (also, Go should not send ANY requests to Sourcegraph.com - you must check the background script's Network tab to confirm)
- Public repos should work as usual (check both the content script and background script for errors you haven't seen before)

Depends on https://github.com/sourcegraph/sourcegraph-langserver-http/pull/8
Depends on https://github.com/sourcegraph/extensions-client-common/pull/57

### CODE REVIEW

File: src/shared/util/dom.tsx [hunk 1]
```diff
@@ -12,3 +12,31 @@ export function insertAfter(el: HTMLElement, referenceNode: Node): void {
 export function isMouseEventWithModifierKey(e: MouseEvent): boolean {
     return e.altKey || e.shiftKey || e.ctrlKey || e.metaKey || e.which === 2
 }
+
+export function commitIDFromPermalink({
```
Comment (ijsnow) [inline src/shared/util/dom.tsx]: *No action required for this PR*

This looks like it could be abstracted slightly more and used for other code hosts if we add a parameter `container` which defaults to `document`. In other code hosts, the commit is taken per code view, not per page(though most of the time each code view will be at the same commit for any page you're looking at).

File: src/shared/util/dom.tsx [hunk 2]
```diff
@@ -12,3 +12,31 @@ export function insertAfter(el: HTMLElement, referenceNode: Node): void {
...
+    selector,
+    hrefRegex,
+}: {
+    selector: string
+    hrefRegex: RegExp
+}): string | Error {
+    const permalinkElement = document.querySelector(selector)
```
Comment (ijsnow) [inline src/shared/util/dom.tsx]: Based on this functons name, I would consider casting this to a link and then below we don't have to use `getAttribute`. 

```suggestion
    const permalinkElement = document.querySelector<HTMLAnchorElement>(selector)
```

File: src/shared/util/dom.tsx [hunk 3]
```diff
@@ -12,3 +12,31 @@ export function insertAfter(el: HTMLElement, referenceNode: Node): void {
...
+    if (!permalinkElement) {
+        return new Error(
+            `Unable to determine the commit ID (40 character hash) you're on because the permalink shortcut element (query selector ${selector}) which contains the commit ID does not exist in the DOM.`
+        )
+    }
+    const href = permalinkElement.getAttribute('href')
```
Comment (ijsnow) [inline src/shared/util/dom.tsx]: If you accept the above suggestions, you can directly get the `href` property and not need to use `getAttribute`.

```suggestion
    const href = permalinkElement.href
```

File: src/shared/util/dom.tsx [hunk 4]
```diff
@@ -12,3 +12,31 @@ export function insertAfter(el: HTMLElement, referenceNode: Node): void {
...
+    if (!href) {
+        return new Error(
```
Comment (ijsnow) [inline src/shared/util/dom.tsx]: Throwing an error is more common in JS

File: src/libs/github/file_info.ts [hunk 5]
```diff
@@ -64,21 +66,36 @@ export const resolveDiffFileInfo = (codeView: HTMLElement): Observable<FileInfo>
...
         )
-    )
+    }
+
+    const commitID = commitIDFromPermalink({
+        selector: '.js-permalink-shortcut',
+        hrefRegex: new RegExp('^/.*?/.*?/blob/([0-9a-f]{40})/'),
+    })
+
+    if (isErrorLike(commitID)) {
```
Comment (ijsnow) [inline src/libs/github/file_info.ts]: Throwing the error inside `commitIDFromPermalink` would prevent the need for this, but you could add a check for whether its not a string here to prevent the need for the separate declaration below

```suggestion
    if (isErrorLike(commitID) || typeof commitID !== 'string') {
```

File: src/libs/gitlab/file_info.ts [hunk 6]
```diff
@@ -29,21 +30,33 @@ const ensureRevisionsAreCloned = (files: Observable<FileInfo>): Observable<FileI
...
+        return throwError(
+            new Error(
+                `Unable to determine the file path of the current file because the current URL (window.location ${
+                    window.location
+                }) does not have a file path.`
             )
         )
-    )
+    }
+    const commitID = commitIDFromPermalink({
```
Comment (ijsnow) [inline src/libs/gitlab/file_info.ts]: I purposefully kept DOM stuff outside of this file. Could you move this into `scrape`? Why is this needed if we get the rev from `getFilePageInfo` anyways?

File: src/libs/gitlab/file_info.ts [hunk 7]
```diff
@@ -29,21 +30,33 @@ const ensureRevisionsAreCloned = (files: Observable<FileInfo>): Observable<FileI
...
+        selector: '.js-data-file-blob-permalink-url',
+        hrefRegex: new RegExp('^/.*?/.*?/blob/([0-9a-f]{40})/'),
+    })
+    if (isErrorLike(commitID)) {
```
Comment (ijsnow) [inline src/libs/gitlab/file_info.ts]: Same as the github file
```suggestion
    if (isErrorLike(commitID) || typeof commitID !== 'string') {
```

File: src/libs/gitlab/file_info.ts [hunk 8]
```diff
@@ -29,21 +30,33 @@ const ensureRevisionsAreCloned = (files: Observable<FileInfo>): Observable<FileI
...
+        return throwError(commitID)
+    }
+    // TypeScript doesn't realize that `commitID` is not an `Error` here.
+    const stringCommitID: any = commitID
```
Comment (ijsnow) [inline src/libs/gitlab/file_info.ts]: Same as the github file. 
```suggestion
    const stringCommitID = commitID as string
```

File: src/libs/github/file_info.ts [hunk 9]
```diff
@@ -64,21 +66,36 @@ export const resolveDiffFileInfo = (codeView: HTMLElement): Observable<FileInfo>
...
+        return throwError(commitID)
+    }
+
+    // TypeScript doesn't realize that `commitID` is not an `Error` here.
+    const stringCommitID: any = commitID
+    return of({
+        repoPath,
+        filePath,
+        commitID: stringCommitID,
+        rev: stringCommitID,
```
Comment (ijsnow) [inline src/libs/github/file_info.ts]: This will mean that if someone is looking at a file on a branch, the pretty branch url will be blown away for the commit, which is a regression. I'm not sure where the issue is for this but this was an issue I cause/fixed before. This should use the `rev` from `parseURL`.

File: src/libs/gitlab/file_info.ts [hunk 10]
```diff
@@ -29,21 +30,33 @@ const ensureRevisionsAreCloned = (files: Observable<FileInfo>): Observable<FileI
...
+    return of({
+        repoPath,
+        filePath,
+        commitID: stringCommitID,
+        rev: stringCommitID,
```
Comment (ijsnow) [inline src/libs/gitlab/file_info.ts]: This will mean that if someone is looking at a file on a branch, the pretty branch url will be blown away for the commit, which is a regression. I cause this on GitHub before and had to fix it from user complaints. This should use the `rev` from `getFilePageInfo`.

File: src/shared/backend/graphql.tsx [hunk 11]
```diff
@@ -113,61 +132,10 @@ function shouldResponseTriggerRetryOrError(response: any): boolean {
...
- */
-export function queryGraphQLNoRetry(
-    ctx: RequestContext,
-    query: string,
-    variables: any = {},
-    url: string = sourcegraphUrl
-): Observable<QueryResult<IQuery>> {
-    return requestGraphQL(ctx, query, variables, url, false) as Observable<QueryResult<IQuery>>
-}
+export const queryGraphQL = (args: GraphQLRequestArgs) => requestGraphQL(args) as Observable<QueryResult<IQuery>>
```
Comment (ijsnow) [inline src/shared/backend/graphql.tsx]: Can we use generics to not have to cast with `as` in so many places?

File: src/libs/gitlab/file_info.ts (hunk 6 above)
Comment (chrismwendt) [reply src/libs/gitlab/file_info.ts]: Moved to `scrape`. `getFilePageInfo` doesn't resolve to a commit ID, does it? It looks like it only gives you the revision (i.e. it could be a ref like `master` rather than a commit ID).

**Comment by chrismwendt** [comment]
Thanks, @ijsnow, I accepted all of your suggestions except the `.href` one :sparkles:

> The changes look nice but unnecessary for the changes intended in this PR. Could we break that out separately so we can easily merge our PRs and then get those changes in afterwards?

Mind if I try resolving the conflicts? No conflict is too difficult to surmount with the almighty [`diffme`](https://gist.github.com/chrismwendt/d2fed14dbb2c003bfb823005d9d3f8fe) 🔨

**Comment by chrismwendt** [comment]
I resolved the conflicts in `src/shared/backend/graphql.tsx` and `src/shared/backend/server.ts` (there didn't appear to be any fundamental conflicts, just some code shuffling).

File: src/shared/util/dom.tsx (hunk 3 above)
Comment (ijsnow) [reply src/shared/util/dom.tsx]: I would prefer `.href` becuase [I use it elsewhere](https://sourcegraph.com/github.com/sourcegraph/browser-extensions/-/blob/src/libs/gitlab/scrape.ts#L160) but its not worth spending much more time on. I'm fine with `getAtrribute`.
